File: fruit-into-baskets/fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int max = -1;
        int n1 = fruits[0];
        int n2 = -1;
        int count = 1;
        int pos = 0;
        for(int i=1; i<fruits.length; i++) {
            if(fruits[i] == n1) {
                count++;
            } else if(n2 == -1) {
                n2 = fruits[i];
                count++;
            } else if(fruits[i] == n2) {
                count++;
            } else {
                max = Math.max(max, count);
                count = i - pos + 1;
                n1 = fruits[i-1];
                n2 = fruits[i];
            }
            if(fruits[i] != fruits[i-1])
                pos = i;
        }
        max = Math.max(max, count);
        return max;
    }
}

File: 940-fruit-into-baskets/fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int l=0, r=1, n=fruits.size(), c1=fruits[0], c2=-1, ans=1, l1=0;
        while(r<n){
            if(fruits[r]!=fruits[r-1]){
                if(c2==-1){
                    c2=fruits[r];
                }
                else if(fruits[r]!=c1 && fruits[r]!=c2){
                    c1=fruits[r-1];
                    c2=fruits[r];
                    ans=max(ans, r-l);
                    l=l1;
                }
                l1=r;
            }
            r++;
        }
        return max(ans, r-l);
    }
};